src/git/repo.rs

```diff
@@ -1446,6 +1711,152 @@ impl<'repo> Branch<'repo> {
     }
 }
 
+/// A Git file status indicator.
+/// See https://git-scm.com/docs/git-status#_short_format
+#[allow(missing_docs)]
+#[derive(Debug, PartialEq, Eq)]
+pub enum FileStatus {
+    Unmodified,
+    Modified,
+    Added,
+    Deleted,
+    Renamed,
+    Copied,
+    Unmerged,
+    Untracked,
+    Ignored,
+}
+
+/// The status of a file in the repo.
+#[derive(Debug, PartialEq, Eq)]
+pub struct StatusEntry {
+    /// The status of the file in the index.
+    pub index_status: FileStatus,
+    /// The status of the file in the working copy.
+    pub working_copy_status: FileStatus,
+    /// The file mode of the file in the working copy.
+    pub working_copy_file_mode: FileMode,
+    /// The file path.
+    pub path: PathBuf,
+    /// The original path of the file (for renamed files).
+    pub orig_path: Option<PathBuf>,
+}
+
+impl StatusEntry {
+    /// Returns the paths associated with the status entry.
+    pub fn paths(&self) -> Vec<PathBuf> {
+        let mut result = vec![self.path.clone()];
+        if let Some(orig_path) = &self.orig_path {
+            result.push(orig_path.clone());
+        }
+        result
+    }
+}
+
+fn parse_file_status(status: u8) -> eyre::Result<FileStatus> {
+    let status = match status {
+        b'.' => FileStatus::Unmodified,
+        b'M' => FileStatus::Modified,
+        b'A' => FileStatus::Added,
+        b'D' => FileStatus::Deleted,
+        b'R' => FileStatus::Renamed,
+        b'C' => FileStatus::Copied,
+        b'U' => FileStatus::Unmerged,
+        b'?' => FileStatus::Untracked,
+        b'!' => FileStatus::Ignored,
+        _ => {
+            warn!(?status, "invalid status indicator");
+            FileStatus::Untracked
+        }
+    };
+    Ok(status)
+}
+
+// Parses the string representation of a filemode for a status entry.
+// Git only supports a small subset of Unix octal file mode permissions.
+// See http://git-scm.com/book/en/v2/Git-Internals-Git-Objects
+fn parse_file_mode(file_mode: &str) -> eyre::Result<FileMode> {
```

Comment:
Likewise. There's also a `FromStr` trait you could increment for specifically the case of converting from `&str`.

Reply:
Done, though I had to create a wrapper type for `FileMode` (which, is probably for the best anyways, to prevent the proliferation of `git2` types)